db/dcrpg/internal/addrstmts.go

```diff
@@ -225,6 +226,15 @@ const (
 	SetTxTypeOnAddressesByVinAndVoutIDs = `UPDATE addresses SET tx_type=$1 WHERE
 		tx_vin_vout_row_id=$2 AND is_funding=$3;`
 
+	UpdateMatchingFundingTxIdFromVins = `UPDATE addresses SET matching_tx_index=tx_index
+		FROM vins WHERE matching_tx_hash=vins.tx_hash AND addresses.tx_hash=prev_tx_hash
+		AND addresses.tx_vin_vout_index=prev_tx_index AND matching_tx_hash !=''
+		AND is_funding=true AND addresses.id >= $1 AND addresses.id < $2;`
+
+	UpdateMatchingSpendingTxIdFromVins = `UPDATE addresses SET matching_tx_index=vins.prev_tx_index
+		FROM vins WHERE vins.id=tx_vin_vout_row_id AND is_funding=false AND matching_tx_hash != ''
+		AND addresses.id >= $1 AND addresses.id < $2;`
+
```

Comment:
@chappjc and @dmigwi  these are the two upgrade queries.  I would appreciate your feedback on them.  On my laptop the upgrade process takes 5 hours - I'm guessing on one of your machines it would be faster.  I saw @dmigwi added temporary indexes for one of his upgrades - perhaps I could experiment with that.  

The linkages seem to be accurate and the queries appropriately fill in all lines in address table with non blank matching_tx lines with a matching_tx_index.   However, as with the other queries on this table I'm not sure if we still have ambiguity due to the sidechains or re-mined items. 

Your feedback would be appreciated.

Reply:
Rerunning the update for `is_funding=TRUE` with `EXPLAIN ANALYZE`:

```
                                                                        QUERY PLAN                                                              
          
------------------------------------------------------------------------------------------------------------------------------------------------
----------
 Update on addresses  (cost=5816983.24..9533917.15 rows=165117827 width=217) (actual time=1842016.473..1842016.473 rows=0 loops=1)
   ->  Merge Join  (cost=5816983.24..9533917.15 rows=165117827 width=217) (actual time=26565.186..146692.090 rows=10657405 loops=1)
         Merge Cond: ((vins.prev_tx_hash = addresses.tx_hash) AND (vins.prev_tx_index = addresses.tx_vin_vout_index))
         ->  Index Scan using uix_vin_prevout on vins  (cost=0.56..839737.22 rows=12261598 width=83) (actual time=0.009..91095.270 rows=12261535
 loops=1)
         ->  Materialize  (cost=5697774.36..5788845.58 rows=18214243 width=207) (actual time=24377.136..45082.620 rows=13161750 loops=1)
               ->  Sort  (cost=5697774.36..5743309.97 rows=18214243 width=207) (actual time=24377.134..43378.516 rows=13161719 loops=1)
                     Sort Key: addresses.tx_hash, addresses.tx_vin_vout_index
                     Sort Method: external merge  Disk: 2780144kB
                     ->  Seq Scan on addresses  (cost=0.00..1386348.06 rows=18214243 width=207) (actual time=0.028..15872.825 rows=13161720 loop
s=1)
                           Filter: is_funding
                           Rows Removed by Filter: 10657385
 Planning time: 2.493 ms
 Execution time: 1842124.734 ms
(13 rows)
```

On my machine that's a little more than 30 minutes.

The large disk usage on the sort on (addresses.tx_hash, addresses.tx_vin_vout_index) is damn large (2,780,144 kB).  Honestly I don't know what PG is doing this.  I'm wondering if it might help to do a subquery to just get (id, tx_hash, tx_vin_vout_index) in a temp table seeing as it's doing a materialize anyway.

Reply:
It doesn't do the sort and materialize using all that disk space when actually doing a batch of id ranges as would be done for real, so never mind the last point.

Reply:
Anyway, I only removed what seemed like unnecessary conditionals from your upgrade queries and it came out like this:

```
2018-09-17 16:57:53.220 [INF] PSQL: Set the matching tx index in address table...
2018-09-17 16:57:53.221 [INF] PSQL: Updating address table entries 0 to 100000 of 23819112
2018-09-17 16:58:07.900 [INF] PSQL: Updating address table entries 100000 to 200000 of 23819112
2018-09-17 16:58:27.183 [INF] PSQL: Updating address table entries 200000 to 300000 of 23819112
2018-09-17 16:58:41.982 [INF] PSQL: Updating address table entries 300000 to 400000 of 23819112
2018-09-17 16:58:52.832 [INF] PSQL: Updating address table entries 400000 to 500000 of 23819112
2018-09-17 16:59:07.257 [INF] PSQL: Updating address table entries 500000 to 600000 of 23819112
2018-09-17 16:59:25.228 [INF] PSQL: Updating address table entries 600000 to 700000 of 23819112
2018-09-17 17:31:39.985 [INF] PSQL: Updating address table entries 23800000 to 23900000 of 23819112
2018-09-17 17:31:41.570 [INF] PSQL:  - 21314759 rows in addresses table (matching tx index column upgrade) were successfully upgraded in 2028.35 sec
```

So about 34 minutes.  Not so bad.

Reply:
Just this diff:

```diff
@@ -227,12 +227,12 @@ const (
                tx_vin_vout_row_id=$2 AND is_funding=$3;`
 
        UpdateMatchingFundingTxIdFromVins = `UPDATE addresses SET matching_tx_index=tx_index
-               FROM vins WHERE matching_tx_hash=vins.tx_hash AND addresses.tx_hash=prev_tx_hash
-               AND addresses.tx_vin_vout_index=prev_tx_index AND matching_tx_hash !=''
+               FROM vins WHERE addresses.tx_hash=prev_tx_hash
+               AND addresses.tx_vin_vout_index=prev_tx_index
                AND is_funding=true AND addresses.id >= $1 AND addresses.id < $2;`
 
        UpdateMatchingSpendingTxIdFromVins = `UPDATE addresses SET matching_tx_index=vins.prev_tx_index
-               FROM vins WHERE vins.id=tx_vin_vout_row_id AND is_funding=false AND matching_tx_hash != ''
+               FROM vins WHERE vins.id=tx_vin_vout_row_id AND is_funding=false
                AND addresses.id >= $1 AND addresses.id < $2;`
 
        IndexBlockTimeOnTableAddress   = `CREATE INDEX block_time_index ON addresses (block_time);`
```

Reply:
Actually, for `UpdateMatchingFundingTxIdFromVins`, we'll also want to add:

`AND vins.is_valid=addresses.valid_mainchain`

The most important rows to update are vins.is_valid=true.  For funding addresses table rows where we have to look into vins table via the prev_tx_hash:prev_tx_index, there may be multiple matching vins rows.  Matching the vins row with is_valid=true is most important, but if there is no vins row with is_valid=true (just 1+ with is_valid=false) and addresses.valid_mainchain=false, we at also want to patch the matching_tx_index to something valid even if there are multiple matching vins rows.